### PR TITLE
Adding Travel Time Distance Validator implmentation based on Google Map API

### PR DESCRIPTION
**Pull request by luchengqi7**
A new implementation of TravelTimeDistanceValidator based on google map API is now added to the matsim-lib. Now users can choose to use either the HERE API or Google Map API.

### CODE REVIEW

**Comment by luchengqi7** [comment]
I will update the previous PR and then merge it to this PR. But @rakow, you can already start reviewing this new PR.

File: contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import java.io.InputStreamReader;
+import java.net.URL;
+import java.text.DecimalFormat;
+import java.text.NumberFormat;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.Locale;
+
+public class GoogleMapRouteValidator implements TravelTimeDistanceValidator {
```
Comment (rakow) [inline contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java:25]: Please add a short small javadoc to public classes.

File: contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        df.applyPattern(pattern);
+
+        String urlString = "https://maps.googleapis.com/maps/api/distancematrix/json" +
+                "?departure_time=" + Long.toString((long) adjustedDepartureTime) +
+                "&destinations=" + df.format(to.getY()) + "%2C" + df.format(to.getX()) +
+                "&origins=" + df.format(from.getY()) + "%2C" + df.format(from.getX()) +
+                "&key=" + apiAccessKey;
+
+        try {
+            System.out.println(urlString);
```
Comment (rakow) [inline contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java:75]: Use logging here

File: contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+                    JSONObject distanceResults = (JSONObject) results.get("distance");
+                    distance = (long) distanceResults.get("value");
+                }
+                if (results.containsKey("duration_in_traffic")) {
+                    JSONObject timeResults = (JSONObject) results.get("duration_in_traffic");
+                    travelTime = (long) timeResults.get("value");
+                }
+            }
+        } catch (IOException | ParseException e) {
+            e.printStackTrace();
```
Comment (rakow) [inline contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java:94]: Also here use logger.

File: contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        String urlString = "https://maps.googleapis.com/maps/api/distancematrix/json" +
+                "?departure_time=" + Long.toString((long) adjustedDepartureTime) +
+                "&destinations=" + df.format(to.getY()) + "%2C" + df.format(to.getX()) +
+                "&origins=" + df.format(from.getY()) + "%2C" + df.format(from.getX()) +
+                "&key=" + apiAccessKey;
+
+        try {
+            System.out.println(urlString);
+            URL url = new URL(urlString);
+            BufferedReader in = new BufferedReader(new InputStreamReader(url.openStream()));
```
Comment (rakow) [inline contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java:77]: It looks like this Reader is not closed. Better wrap it in try-with-resource block.

File: contribs/application/src/main/java/org/matsim/application/analysis/TravelTimeAnalysis.java [hunk 5]
```diff
@@ -39,13 +39,19 @@
...
     @CommandLine.Option(names = "--run-id", defaultValue = "*", description = "Pattern used to match runId", required = true)
     private String runId;
 
-    @CommandLine.Option(names = "--output", defaultValue = "travelTimeResults", description = "Name of output folder", required = true)
+    @CommandLine.Option(names = "--output", defaultValue = "travelTimeValidation", description = "Name of output folder", required = true)
     private String output;
 
-    @CommandLine.Option(names = "--api-key", description = "HERE Maps API key, see here.com", required = true)
+    @CommandLine.Option(names = "--api", description = "Online API used. Choose from [here, google-map (todo)]", defaultValue = "here", required = true)
+    private String api;
```
Comment (rakow) [inline contribs/application/src/main/java/org/matsim/application/analysis/TravelTimeAnalysis.java:52]: You may also add an enum for this.

**Review by rakow — CHANGES_REQUESTED** [review]
This PR has a huge overlap this the previous one. Next time I would suggest not creating a new PR and simply appending to the existing one.

**Comment by luchengqi7** [comment]
> This PR has a huge overlap this the previous one. Next time I would suggest not creating a new PR and simply appending to the existing one.

Yeah, but the change in this PR are partially based on previous PR. Is there any better way to do this in the future?

File: contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java (hunk 4 above)
Comment (rakow) [reply contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java]: Yes.

Comment (luchengqi7) [reply contribs/analysis/src/main/java/org/matsim/contrib/analysis/vsp/traveltimedistance/GoogleMapRouteValidator.java]: Okay. Thanks!

**Review by rakow — APPROVED** [review]

**Pull request merged**